Towards a semi-performant recursive interpreter

timholy opened the pull request:
This is a pretty big overhaul of this package. I've tried to keep its original functionality, but perhaps a split should be the next step (see #32). The first 5 commits get the tests passing, with the exception of the ui tests which I didn't even look at (CC @staticfloat re #34, #36; @pfitzseb re #33).

The rest is much more ambitious. I worked a lot on the performance, since an "easy" way to add robust breakpoints, etc, is via running your code in the interpreter. And of course there's interest in using the interpreter to circumvent compile-time cost. But then performance matters.

The most important changes here are focused on reducing the cost of dynamic dispatch (or its equivalent here, "dynamic lowered code lookup"). Some highlights:
- At the lowest level, all calls are to Builtins and IntrinsicFunctions. I used the `tfunc` data in `Core.Inference` to auto-generate an evaluator that resolve all calls with fixed number of arguments. A straightforward extension would be resolve all calls with bounded numbers of arguments (e.g., those that have between 2 and 4 arguments).
- To reduce the overhead of `which`, this adds "local method tables," one per :call Expr. These are cached by exact type, since `isa(x, Int)` is fast but `isa(x, Integer)` is slow. So it uses `MethodInstance` comparisons rather than `Method` signature comparisons, even though it might look up the same lowered code. I think a slightly more elegant way to do this would be to add a new type,

```julia
struct LocalMethodTable
    call_expr::Expr
    knownmethods::TypeMapEntry
end
```

to the list of valid types in a `CodeInfo`. (I did it this way at first but it breaks things like basic-block computation and ssa-usage analysis. So I resorted to storing this info in a separate fields.)

For a simple summation test, I'm getting about 15us per iteration. Compiled code is about 5ns, so this is still dirt-slow. But just getting it to this point was quite a major overhaul.

CC @JeffBezanson, @vtjnash, @StefanKarpinski.

jpsamaroo commented:
Awesome! What's that summation test look like? The compiled version could possibly be compiling down to a static result (i.e. removing the loop and just returning the result directly), since 5ns is wicked fast. That wouldn't therefore be a good measure of this PR's performance gains.

timholy commented:
The summation test is [here](https://github.com/timholy/ASTInterpreter2.jl/blob/3c4084ec7ff54f6d64171b3ae28bec6262d25ebd/test/interpret.jl#L79-L85). 5ns/iteration is expected for a GHz CPU (it would be faster still if I had added `@inbounds` and `@simd`). So 15us/iteration implies a few thousand CPU/cycles per iteration, which is pretty slow. It would be faster if we didn't care about stashing the result of computations in `JuliaStackFrame`s, but that's essential for this package's utility as a debugger.

timholy commented on src/generate_builtins.jl:
```diff
@@ -0,0 +1,143 @@
+        end
+        print(io,
+"""
+            $head name == :$fname
```
Agreed. I thought of that and then talked myself out of it for fears of introducing the interpreter-equivalent of https://github.com/JuliaLang/julia/issues/265. But now I'm not so sure that's a risk, and in any event I currently clear the cache in between `@interpret` invocations.

KristofferC commented:
Thanks a lot for this. I pushed a commit to update the Project + Manifest to get CI running (hope you don't mind).

KristofferC commented:
So the UI tests pass as long as there is a good way to set the `fullpath` variable to false in the `JuliaFrameCode`. Previously, this was done by a keyword to the `JuliaStackFrame`. Of course I could hack something in but do you have any preference how to expose this option @timholy, or perhaps we could tweak the tests themselves?

timholy commented on Manifest.toml:
```diff
@@ -7,10 +7,18 @@ uuid = "2a0f44e3-6c83-55bd-87e4-b1978d98bd5f"
 uuid = "67417a49-6d77-5db2-98c7-c13144130cd2"
 version = "0.1.2+"
 
+[[DebuggingUtilities]]
```
Sorry, I added this while I was debugging my own work on this PR---`@showln` is very useful. But we don't need it to be part of this package long-term. I'll trust it's OK with you if I overwrite this.

KristofferC replied:
Of course :)

vtjnash commented:
> to auto-generate an evaluator that resolve all calls with fixed number of arguments

That all sounds very confusing. Any reason not to just execute them directly? They are just normal methods, albeit without source code accessible to reflection (and that block overloading).

timholy commented:
OK, this fixes what I think are the remaining problems. This enhances the performance on my loop tests a teeny bit more (down to 13us per iteration), presumably because of resolving `GlobalRef`s during `optimize!`. I also fixed a breakage on master due to the very nice https://github.com/JuliaLang/julia/pull/30641.

The most important new commit is 19c1492cdfdc63cafbe06f1122153a596d25745b, which fixed a serious bug that caused it to sometimes look up the wrong lowered code for generated functions. Now this seems to behave quite robustly; for example, I can run the subarray tests. They're not fast, but not much slower than in compiled mode (). Most of the time is spent on [one line](https://github.com/timholy/ASTInterpreter2.jl/blob/5a78ad4ea8e71f46cc68510c39fd8a56d51236c1/src/interpret.jl#L145) to `eval` `%new` expressions.

@KristofferC, I also added kwargs support back to the `JuliaStackFrame` constructor, so hopefully you can get the ui tests passing. Thanks for tackling that! :heart:

I think this can be merged. (I can't do that, however.) And I do think we should plan on splitting out the DebuggerFramework functionality out very shortly afterwards. There's plenty of reason to be interested in an interpreter independent of its usage for a IDE or REPL debugger. So perhaps before hitting merge we should think a bit about what that will look like (and what the package names should be).

KristofferC commented:
One thing that I know have been discussed is to rename this package (or whatever is the package that a user will finally interact with). In 0.6 Gallium took that place (simply reexporting ASTInterpreter2) but I wonder if we should just come up with something fresh.

timholy commented:
> That all sounds very confusing. Any reason not to just execute them directly? They are just normal methods, albeit without source code accessible to reflection (and that block overloading).

That is what it does, see https://github.com/timholy/ASTInterpreter2.jl/blob/teh/localmt/src/builtins.jl. Rather than typing those all out, though, I just check the `tfunc` tables in Core.Compiler to see how many args are supported. They are generated by code in https://github.com/timholy/ASTInterpreter2.jl/blob/teh/localmt/src/generate_builtins.jl, which is about half the size of the finished product, and would have the bonus that if any of the builtins or intrinsics ever change, this should pretty much auto-update.

One thing I didn't do: we might want to consider adding `if nargs == ` cases for `arrayref` and `arrayset` for low dimensionality.

timholy commented:
I've thought about `CodeInfoInterpreter.jl` or `LoweredInterpreter.jl` or just `Interpreter.jl`.

Keno commented:
Awesome work. I'll let @KristofferC take this forward in detail. My only concern is that ASTInterpreter2 was supposed to be very small and maintainable, so I'm afraid to add too much here that would make it more complicated. On the other hand, perhaps that's offset by more people looking at that. Also seconded on @vtjnash's question why generating the source file is necessary.

timholy commented on src/builtins.jl:
```diff
@@ -0,0 +1,276 @@
+        return Some{Any}(Core._apply_pure(getargs(args, frame)...))
+    elseif f === Core._expr
+        return Some{Any}(Core._expr(getargs(args, frame)...))
+    elseif f === Core._typevar
```
Looks like this isn't present on 1.0. One wonders if we should avoid committing `builtins.jl` and just generated it during `Pkg.build`?

timholy commented:
> Also seconded on @vtjnash's question why generating the source file is necessary.

We don't have to do it that way, although `Core._typevar` points out some potential advantages to autogeneration (if it's sufficiently robust...you could imagine it going either way as far as maintainability goes).

timholy commented:
> My only concern is that ASTInterpreter2 was supposed to be very small and maintainable, so I'm afraid to add too much here that would make it more complicated. On the other hand, perhaps that's offset by more people looking at that.

I think the biggest reason to do it is that if we can get reasonable performance with the interpreter it becomes feasible to support breakpoints in the short term. (Supporting breakpoints is easy if all the code is running in the interpreter.)

I don't think even a loving author could say that the performance of this PR makes it "reasonable," but it is a ~60x improvement over where I started. If I could get another 10x I'd be really happy. But I'm skeptical that we can get there without some additional higher-level analysis, most important probably being a certain amount of lowered-code inlining to reduce the number of created stackframes. (That seems likely to require a limited form of type-inference...) Julia's polymorphism makes it much more difficult to write a fast interpreter for than, say, languages where `for i = 1:100` is a construct of the language itself (e.g., https://dzone.com/articles/adventures-in-jit-compilation-part-1-an-interptete, where even "simpleinterp" is about 180x faster than this). We have to recurse into every call to `iterate`.

vtjnash commented:
> Most of the time is spent on one line to eval %new expressions.

It'd probably be much faster to ccall that function directly, rather than having `eval` do it. I've seen calls to that ccall elsewhere in this PR already...

> Rather than typing those all out

That's good to hear, since that list is large, but `f isa Core.Intrinsic && f(args...)` should run faster than checking for the values 1 to 100 in separate if statements. If you want to skip `_apply`, hard coding for length instead should be fast and general (e.g. unroll each call site for 1-N args, where N is a small integer of about 6)

timholy commented:
Doesn't actually work out that way:

```julia
julia> using BenchmarkTools, ASTInterpreter2

julia> function runifbuiltin(qf, qx)
           f, x = qf.value, qx.value
           if isa(f, Core.Builtin)
               return Some{Any}(f(x))
           end
           return qx
       end
runifbuiltin (generic function with 1 method)

julia> function runifintrinsic(qf, qx)
           f, x = qf.value, qx.value
           if isa(f, Core.IntrinsicFunction)
               return Some{Any}(f(x))
           end
           return qx
       end
runifintrinsic (generic function with 1 method)

julia> qf = QuoteNode(abs)
:($(QuoteNode(abs)))

julia> qx = QuoteNode(3)
:($(QuoteNode(3)))

julia> @btime runifbuiltin(qf, qx)
  64.096 ns (0 allocations: 0 bytes)
:($(QuoteNode(3)))

julia> @btime runifintrinsic(qf, qx)
  11.302 ns (0 allocations: 0 bytes)
:($(QuoteNode(3)))

julia> ex = Expr(:call, qf, qx)
:(($(QuoteNode(abs)))($(QuoteNode(3))))

julia> @btime ASTInterpreter2.maybe_evaluate_builtin(nothing, ex)
  16.826 ns (0 allocations: 0 bytes)
:(($(QuoteNode(abs)))($(QuoteNode(3))))

julia> qf = QuoteNode(Core.sizeof)
:($(QuoteNode(Core.sizeof)))

julia> @btime runifbuiltin(qf, qx)
  96.828 ns (1 allocation: 16 bytes)
Some(8)

julia> @btime runifintrinsic(qf, qx)
  14.321 ns (0 allocations: 0 bytes)
:($(QuoteNode(3)))

julia> ex = Expr(:call, qf, qx)
:(($(QuoteNode(Core.sizeof)))($(QuoteNode(3))))

julia> @btime ASTInterpreter2.maybe_evaluate_builtin(nothing, ex)
  30.768 ns (1 allocation: 16 bytes)
Some(8)

julia> qf = QuoteNode(Base.neg_int)
:($(QuoteNode(neg_int)))

julia> @btime runifbuiltin(qf, qx)
  113.218 ns (1 allocation: 16 bytes)
Some(-3)

julia> @btime runifintrinsic(qf, qx)
  56.686 ns (1 allocation: 16 bytes)
Some(-3)

julia> ex = Expr(:call, qf, qx)
:(($(QuoteNode(neg_int)))($(QuoteNode(3))))

julia> @btime ASTInterpreter2.maybe_evaluate_builtin(nothing, ex)
  45.757 ns (1 allocation: 16 bytes)
Some(-3)
```

I've also tried a middle ground, checking `isa(f, Core.IntrinsicFunction)` and then "dispatching" by number of arguments, and that's slower. I've not found any approach that's faster than the one in this PR.

timholy commented:
> It'd probably be much faster to ccall that function directly, rather than having eval do it.

Good call! 900s->590s.

KristofferC commented:
I updated the UI tests (and their dependencies) so tests now pass on 1.1 and nightly. 1.0 fails due to the already mentioend `_typevar` issue.

KristofferC commented on Project.toml:
```diff
@@ -6,12 +6,13 @@ version = "0.1.1"
 
 [targets]
 test = ["Test", "TerminalRegressionTests", "VT100"]
+build = ["InteractiveUtils"]
```
There is currently no `build` target.

timholy replied:
Recommended approach? Just make it a dependency of the package?

KristofferC replied:
Yes

timholy commented:
OK, I've modified this to generate the file at build time. I expect tests to pass. Before merging, I think it would be best to wait a bit in case anyone wants to give this a detailed review. 

There are also a couple of issues to discuss. First, I recognize the excitement to get debugging working in Juno again. But I'd urge a bit of patience to tackle a couple of things that should be done first:

- any package splitting/reorganization/renaming
- integration with Revise's ability to correct line numbers. A debugger is nice for, well, fixing bugs. But if this results in line number changes in methods that don't get recompiled, the next time you try to step into code you'll presumably be at the wrong line number of the source file, and users will be really confused.
- a couple of days for me to experiment with this for interpreting code in module-scope. I *think* it should work, but I haven't tried to play with it. Revise will evaluate changing its internal organization to use a stripped-down ASTInterpreter2 (just the interpreter part) to more robustly find method signatures and add "lowering backedges" (see #32).

The point being that I don't think it would be great to reintroduce a stepping-debugger to the Julia world and then break it only a couple of days later.

KristofferC commented:
Hm, it seems test only dependencies that are using a checked out branch in the Manifest aren't too happy when they get put in the `test` section. Will investigate.

timholy commented:
Also one technical issue to highlight: before merging, perhaps I should probably experiment with moving the local method tables from the `JuliaFrameCode` to the `JuliaStackFrame` (the first being the "reusable component" and the second being the "instance for this particular call"). This may have performance advantages, if I can do it in a way that doesn't throw out the method tables for any recursive calls.

timholy commented:
To ensure that this package can be hacked on by more than a small handful of people, I've added a bunch of docstrings and some Documenter narrative, including a somewhat gentle introduction to Julia's lowered AST representation. I haven't implemented deployment yet since I think it would be better to do that once we're basically ready to release. But you can view the docs locally by navigating to the `docs` directory and running `julia make.jl` and then opening `build/index.html` in a browser.

tpapp commented:
Big thanks for the docstrings. I would like to contribute, but I am gaining understanding of the codebase quite slowly, and I think they will help a lot.

timholy commented:
Main thing is to make sure you understand the AST representation---once you have that under control, the rest really isn't terribly complicated, and is mostly just a matter of familiarization. Julia's [AST docs](https://docs.julialang.org/en/latest/devdocs/ast/) are fantastic, but for getting oriented the new `ast.md` page should be useful. Then the `internals.md` page should explain most of the rest of what you need.

Advanced topics include things like kwargs, generated functions, and the storage of local method tables. But I'd urge you to ignore these things at first; just focus on `enter_call` and then `step_expr`. The latter is unfortunately long, but if you shrink your fontsize enough you can use Rebugger to step into it (using one debugger to understand another :smile:), and that may help you understand the control-flow.

timholy commented:
@Keno, one technical question: why are the `ssavalues` offset by one slot? I noticed this especially when writing up the docs; it seems more straightforward to store the result of statement `%2` at `frame.ssavalues[2]` rather than `frame.ssavalues[3]`. Any reason I'm unaware of? (Or is it historical, did SSAs used to start numbering at 0?)

KristofferC commented:
For the tests to pass, I think it is easiest to wait for new tags for the dependent packages and just use these. I pinged Keno so will likely happen today.

timholy commented:
Very good.

Once this merges, what do you thing about the name "DebuggingInterpeter" for a package that weaves together this one and DebuggingFramework. I.e.,

ASTInterpreter2: becomes a "pure interpreter"
DebuggingInterpreter: imports ASTInterpreter2, DebuggingFramework. This package gets all the code here that relies on DebuggingFramework.

Of course we can additionally rename this one.

pfitzseb commented:
`Debugger.jl` seems perfect for the user facing package, especially since we can already implement most features one would expect with the existing functionality (stepping already works, budget breakpoints are easy with the improved performance, and break-on-error could maybe reuse Rebugger.jl for now).
If some (e.g. Keno) has time to work on a "real" debugger (like Gallium) then we can switch `Debugger.jl`s internals anyways.

So my suggestion would be:
- `Interpreter.jl` as the pure interpreter
- `DebuggingFramework.jl` stays as-is
- `Debugger.jl` as the frontend

timholy commented:
OK. 

> We could piggy-back on the Gallium name

I think we should really reserve that for something that can directly introspect call stacks. The x86 trap instruction `0x1f` and Gallium's atomic number `31 == 0x1f` is a pun that's too clever to waste.

One more change I should have mentioned: in my (latest, barely-started) rewrite of Revise, I will split it into two packages. One package, perhaps called `CodeTracking.jl`, will be a minimal dependency that can be used to ask

```julia
file, lineno = where_is_that_darned_method_now(method)
```

Revise will also import CodeTracking.jl (and probably Interpreter.jl), and will be responsible for doing the work to maintain CodeTracking's global variables. By splitting these apart, it will allow packages like `Debugger.jl` to depend on `CodeTracking` without needing the full weight of Revise itself (or making it required for using these other packages).

So the full dependency chain is:

- `Interpreter.jl`: standalone
- `CodeTracking.jl`: standalone
- `DebuggingFramework.jl`: needs REPL etc but otherwise standalone
- `Breakpoints.jl`: needs `ASTInterpreter.jl`
- `Debugger.jl`: needs `Interpreter.jl`, `CodeTracking.jl`, `DebuggingFramework.jl`, `Breakpoints.jl`
- `Revise.jl`: needs `CodeTracking.jl`, `Interpreter.jl`
- `Juno`: needs `Debugger.jl`
- `Rebugger`: needs `CodeTracking.jl`, `Interpreter.jl`, and `Breakpoints.jl`

Rebugger will depend on time & priorities, i.e., if I and others think it's still useful to have an efficient pure-REPL debugger.  Thanks to Interpreter.jl, it wouldn't be too hard to implement a "step to next line by hitting space, enter into call by hitting Enter, finish by hitting up arrow, interrupt and execute custom statements by hitting Esc" execution mode.)

timholy commented:
I'm not entirely sure what the original purpose of DebuggerFramework was, but I think it was to support implementation-agnostic APIs for debugging.

One thought would be to split DebuggerFramework into "debugging API" and "step/next/continue REPL mode" packages. Does Juno need the latter for anything? I guess to support introspection of the call stack during stepping?

pfitzseb commented:
> One thought would be to split DebuggerFramework into "debugging API" and "step/next/continue REPL mode" packages. Does Juno need the latter for anything? I guess to support introspection of the call stack during stepping?

No, Juno basically implements one and a half more REPL modes for that itself. There are still a couple of functions in Juno that make use of ASTInterpreter right now (getting the expression to be evaluated next, finding the location of a statement, and getting variables in a function local scope), but those would probably end up in another package anyways, I think.

StefanKarpinski commented:
May I propose playing the "briefly describe what it does" game? For each package, try to briefly describe what the package is for and take that phrase into account when naming it. It also helps with the somewhat unfortunate profusion of packages that we seem to have. I know some of them were developed independently—which is totally fine—but a large number of them seem to have been developed together and been split into separate packages rather prematurely and unnecessarily (e.g. DebugFramework). If you can't describe succinctly what a package is for, it's often a hint that it shouldn't exist.

timholy commented:
Good idea, I'll play your game. First the current state:

ASTInterpreter2 combined with DebuggerFramework: an interpreter AND a debugger interface AND and REPL mode.
Revise: diff/patch AND something that presents methods to Rebugger

There's one crucial point that we need to get straight first, concerning the integration of debugging and Revise. Any debugger has to go look up the source code and present it to the user. For example, when you step into `foo`, you'd really like it if Juno opened a new tab and positioned the cursor to the first line of `foo`. So, I'm guessing Juno currently uses the method's `.file` and `.line` fields to know where to navigate in the file. Sounds good, right? But now throw Revise into the mix, suddenly you can be in a situation where the `.file` and `.line` fields are completely useless: suppose the source file containing `foo` also contains, higher up, a 100-line block of comments, and you decide to delete them and then save the file. Note that the starting line of `foo` now decreased by 100, but you didn't need to recompile `foo` because it, itself, didn't change. (And why recompile a method and all of its dependents simply because you want to correct line numbers?) So now when Juno opens the file, it will position the cursor halfway down into a completely different method `bar`; but internally it will still be stepping through `foo`, because Interperter.jl is using the lowered code which is still valid. Super, duper, confusing for users.

Since Revise is tracking the code, the answer is to just ask Revise where the code is now.

So now it's time to play your game, with respect to the new state. Some is quite clear:
- Interpreter.jl: run lowered code without compilation
- Breakpoints: set & remove breakpoints (logical core, not GUI implementation), run lowered code until you hit a breakpoint.
- CodeTracking: the "query" portion of what's currently Revise: where is my method now?
- Revise: will be boiled down to its diff/patch functionality. No longer a dependency of Rebugger, though of course it will usually run in parallel.

The rest is more hazy. Even Breakpoints.jl is "optional" as to whether it's a separate package. In practice this package would essentially be adding a `breakpoints` field to [this structure](https://github.com/timholy/ASTInterpreter2.jl/blob/62adc7234cfcc6dfe615c87857dcc3a5bede330c/src/ASTInterpreter2.jl#L83-L95). One could make the argument that this does not have to be a separate package, but could be integrated into Interpreter.jl. Depends on whether you feel that a notion of breakpoints belongs in something that is nominally "just an interpreter."

oxinabox commented:
I am playing around with my own Cassette based debugger designs.
Obviously there are pro, and cons to this. But not purpose of this discussion. (It is a lot of fun and I am learning a lot about IR, so useful)


Anyway  when I look at the breakdown of packages,
I want to see which bits I can reuse.
If there is absolutely no why I can reuse it then that is an argument against splitting it up  (though not a diffinitive one)

 -  **CodeTracking**.jl:  definately useful. I already using Revise for some of this.
 -  **Breakpoints**.jl: not useful at all,  because the fundermental machanism for setting and hitting breakpoints is different. I can't even picture a way to abstract this notion of a breakpoint away.
 - **DebuggingFramework**.jl: In the ideal world useful. I am not sure that right now its abstractions are abstract enough that I one could use  ASTInterpeter 2or my Cassette based work, but I can *imagine* a world in which they are.
 - **Debugger**.jl: Seems like this just glues DebuggingFramework to ASTInterpeter2? Thus in a world which I can use DebuggingFramework, it is useful becuase it lets that seperation exists

Keno commented:
Happy to leave the organization up to you guys, but couple of points:
- The reason DebuggerFramework.jl exists was because it at the time also support a C/C++ debugger through the same interface, so I tried to keep julia-specific assumptions out of that package
- I usually recommend avoiding splitting packages too much early on as it splits focus.

Keno commented:
> (Or is it historical, 

yes

> did SSAs used to start numbering at 0?)

yes

Keno commented:
Which is to say, just change that ! ;)

timholy commented:
Interesting Pkg failure here.

KristofferC commented:
Sorry, my last commit was bad. Will fix.

timholy commented:
I'm fine with merging this. Interestingly even though I think I now have write access, GitHub doesn't show me the button. (Maybe because I didn't when I first submitted this?) Can you merge? I've worked to keep the commits bisectable, so please don't squash. (Maybe your last couple of commits should be tweaked a bit though.)

KristofferC commented:
Yes, I'll fix up the latter commits and merge.
